public_html/wp-content/mu-plugins/quickbooks/includes/client.php

```diff
@@ -0,0 +1,382 @@
+	 */
+	const OAUTH_REDIRECT_URI = 'https://wordcamp.org/wp-admin/admin-post.php?action=wordcamp-qbo-oauth&cmd=exchange';
+
+	/**
+	 * @var WP_Error|null
+	 */
+	public $error = null;
+
+	/**
+	 * @var DataService|null
+	 */
+	protected $data_service = null;
+
+	/**
+	 * Client constructor.
+	 */
+	public function __construct() {
+		$this->error = new WP_Error();
+
+		if ( ! class_exists( '\QuickBooksOnline\API\DataService\DataService' ) ) {
+			$this->error->add(
+				'missing_dependency',
+				'The required library <code>QuickBooks V3 PHP SDK</code> is unavailable.'
+			);
+
+			return;
+		}
+
+		$config = $this->get_client_config();
+
+		if ( empty( $config['ClientID'] ) || empty( $config['ClientSecret'] ) ) {
+			$this->error->add(
+				'missing_credentials',
+				'The required credentials for connecting to the QBO API are unavailable.'
+			);
+
+			return;
+		}
+
+		try {
+			$this->data_service = DataService::Configure( $config );
+		} catch ( SdkException $exception ) {
+			$this->add_error_from_exception( $exception );
+		}
+
+		if ( isset( $config['refreshTokenKey'] ) ) {
+			try {
+				/**
+				 * The OAuth access token expires after an hour, so whenever we instantiate with saved token data, we
+				 * need to refresh it with the refresh token to get a current access token. According to the SDK docs,
+				 * the refresh token may change when this happens, so it needs to be re-saved during this process.
+				 *
+				 * See https://intuit.github.io/QuickBooks-V3-PHP-SDK/authorization.html#oauth-2-0-vs-1-0a-in-quickbooks-online
+				 */
+				$token = $this->data_service->getOAuth2LoginHelper()->refreshToken();
+				self::save_oauth_token_data( $token );
+			} catch ( SdkException | ServiceException $exception ) {
+				$this->add_error_from_exception( $exception );
+
+				if ( $this->error->get_error_messages( 'invalid_grant' ) ) {
+					$this->error->add(
+						'broken_oauth_connection',
+						'The connection to QuickBooks has failed. Please try reconnecting.',
+						$this->error->get_error_messages( 'invalid_grant' )
+					);
+
+					$this->error->remove( 'invalid_grant' );
+
+					// The bad token data needs to be removed or the SDK will throw an exception when trying to reconnect.
+					self::delete_oauth_token_data();
+				}
+			}
+		}
+	}
+
+	/**
+	 * Does the client have access to the SDK?
+	 *
+	 * @return bool
+	 */
+	public function has_sdk() {
+		return $this->data_service instanceof DataService;
+	}
+
+	/**
+	 * Configuration parameters for the SDK's DataService class.
+	 *
+	 * @return array
+	 */
+	protected function get_client_config() {
+		/**
+		 * Filter: Add configuration values for the QBO V3 PHP SDK.
+		 *
+		 * @param array $config {
+		 *     See https://intuit.github.io/QuickBooks-V3-PHP-SDK/configuration.html
+		 *
+		 *     @type string $auth_mode
+		 *     @type string RedirectURI
+		 *     @type string $scope
+		 *     @type string $ClientID
+		 *     @type string $ClientSecret
+		 *     @type string $baseUrl
```

Comment:
It might be worth adding some short docs for each param here, just in case the above link changes or goes away. Especially for `scope` & `auth_mode`, the others are more self-explanatory.

Reply:
Good idea. Added: 3b8ab198